src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs

```diff
@@ -1177,6 +1175,40 @@ private void setMethodAttribs(CORINFO_METHOD_STRUCT_* ftn, CorInfoMethodRuntimeF
             // TODO: Inlining
         }
 
+        private bool canTailCall(CORINFO_METHOD_STRUCT_* callerHnd, CORINFO_METHOD_STRUCT_* declaredCalleeHnd, CORINFO_METHOD_STRUCT_* exactCalleeHnd, bool fIsTailPrefix)
+        {
+            if (!fIsTailPrefix)
+            {
+                MethodDesc caller = HandleToObject(callerHnd);
+
+                // Do not tailcall out of the entry point as it results in a confusing debugger experience.
+                if (caller is EcmaMethod em && em.Module.EntryPoint == caller)
```

Comment:
Could we return `true` here and in the last check if `StackTraceSupport` is disabled? Same question for that `CORINFO_FLG_DONT_INLINE_CALLER` flag above that's not used on AOT too: could we not skip that too when `StackTraceSupport` is disabled, so we never miss optimizations there if we don't care about the stack trace anyway?

Reply:
> Could we return `true` here and in the last check if `StackTraceSupport` is disabled? Same question for that `CORINFO_FLG_DONT_INLINE_CALLER` flag above that's not used on AOT too: could we not skip that too when `StackTraceSupport` is disabled, so we never miss optimizations there if we don't care about the stack trace anyway?

What would be the advantage? Conditioning this on StackTraceSupport would only make sense if we also check this is GetCallingAssembly (and not some other method marked DynamicMethod). But in that case this API is going to throw anyway. Do we need it to throw faster?